### ISSUE TITLE
CS2030 Project: Is there a need for more events than whatever was instructed?

### ISSUE THREAD

**Issue by jtcones** [nus-cs2030/2324-s2#473]
# Summary
I am wondering whether it is necessary or avoidable to have more events than what was instructed (Arrive, Wait, Serve, Done, Leave)

## Description 
In my implementation, it seems like having a Rest event was the only way for me to ensure that I can pass on the time to my next event. 

My implementation keeps track of an ImList<ImList<Event>> where each ImList<Event> is the queue of events for each server. 

My Server only have attributes int ID and Supplier<Double> restTimes. 

I pass the time on from one event to another event. 

I required a Rest Event to be in the servers queue to ensure that it stalls the queue in the event that another customer arrives before the Servers rest is done. 

How did you guys do it? Does my implementation encapsulate and maintain abstraction?

**Comment by ZHD1987E** [comment]
It’s allowed. Trust me bro.

**Comment by jtcones** [comment]
> It’s allowed. Trust me bro.

Do you think it follows the principles of encapsulation and abstraction?

**Comment by luckrainbow** [comment]
Not compulsory, though I'm sure it'll be accepted as it is logical.

**Comment by samuelneo** [comment]
I did it without a Rest event, and handled the logic for resting in Done. customers can just rewait if they need to

**Comment by jtcones** [comment]
> I did it without a Rest event, and handled the logic for resting in Done. customers can just rewait if they need to

Does your implementation use a nested ImList too? My logic is if the ImList is empty then the Server is available to serve immediately. 

Should there be another Serve event in queue, I will not create the rest event but straight away update the time of that serve event. 

However should the queue be empty after serving the latest known customer. I require the rest event to let the upcoming customer to know you cant be served immediately. 

I cant seem to find a work around for this

**Comment by ZHD1987E** [comment]
I checked back:
- I literally created 12 Event classes
- 6 of them cater to the customer being enqueued to a self-checkout
- The rest are catered to the human version
- Abstraction and Encapsulation is there to modularise your code.

**Comment by ZHD1987E** [comment]
@jtcones A 'rest' event IS OK! That way you can modularise your code further.

**Comment by samuelneo** [comment]
> Does your implementation use a nested ImList too? My logic is if the ImList is empty then the Server is available to serve immediately.

no, the PQ handles all my events. my Server has an attribute for its next available time, so it can just push that back when it needs to rest

**Comment by ZHD1987E** [comment]
Same as what @samuelneo said.

**Comment by yushanedi** [comment]
@ZHD1987E That's great to hear Hao Dong! Terrific stuff 👍🏻

**Comment by Monoji77** [comment]
I did not use a final intermediate event. If coded correctly, the "rest" can be implemented along with your logic in the done class without breaking OOP. The trick is to consider how you want to structure your servers to hide the "rest" logic from the Done event. All the best :)

**Comment by hohinnn** [comment]
I did not make a "rest" event as well, I felt like it was slightly redundant and put it into my "done" event instead!

**Comment by hadyamanisanto** [comment]
> I did not make a "rest" event as well, I felt like it was slightly redundant and put it into my "done" event instead!

Same here.

**Comment by jaiwee** [comment]
> I did not make a "rest" event as well, I felt like it was slightly redundant and put it into my "done" event instead!

yup me too! i think you can just call a method in server to update the available time, in the done event

**Comment by JovanFoo** [comment]
> I did not make a "rest" event as well, I felt like it was slightly redundant and put it into my "done" event instead!

Same here!

**Comment by kohweien** [comment]
> I did not make a "rest" event as well, I felt like it was slightly redundant and put it into my "done" event instead!

me too! after a customer is served, i had the first done event with a printed output, then i called a second done event and updated the server's next available time without any printed output

**Comment by iceflakes12** [comment]
Usually, server should not know event exists. You do not need an ImList of Events for each server. But depending on how you implement, your way could work also. So try to stick with your current implementation unless you encounter any issues.

Creating additional events is fine, and I find it unlikely that creating additional events would break encapsulation or abstraction. After all you encapsulate values into that additional event and abstract away the details of methods for that event.

**Comment by mChewWW** [comment]
it is up to you to decide how many events/other additional classes you want to make, but trying to condense many functions into one event/class may mean that you are not properly applying the concepts of encapsulation.

**Comment by yishao17** [comment]
I did not use a rest event as opposed to some here, but instead I updated the rest time of server in simulator as it is where the .get() is. I think there are more than 1 method to do it too!

**Comment by zuoshihua** [comment]
> I checked back:
> 
> - I literally created 12 Event classes
> 
> - 6 of them cater to the customer being enqueued to a self-checkout
> 
> - The rest are catered to the human version
> 
> - Abstraction and Encapsulation is there to modularise your code.

So... if there's another type of server you'll need to create another 6 classes?

**Comment by ZHD1987E** [comment]
Hey @zuoshihua!

It actually depends. By the way, ever since I did code review I cut down on the number of classes required. (If I am not wrong I did away 4 classes with some code revision, I'll make my project README public soon).

Whenever a new server type is introduced I only need to create 3 classes: the `Wait`, the `Serve` and the `Done` classes corresponding to that type of server.

**Comment by nicthepurplejelly** [comment]
me too! I created an intermediate event, such that if my servers/counters is not yet available then it will help me update the next available time of my counters/servers until it is free to serve the next customer in the queue.

**Comment by ZHD1987E** [comment]
@jtcones Did you manage to do it?

If so, you might want to close this issue. Thank you!

**Comment by lohkeyyy** [comment]
Personally i did it without any additional events, however, in your case, a Rest event seems appropriate. And to answer your question, yes i think it is allowed and it does not violate any OOP principles.

**Comment by ZHD1987E** [comment]
@jtcones The project is over.
